li-protobuf/src/main/java/com/linkedin/data/protobuf/Utf8Utils.java

```diff
@@ -274,191 +274,320 @@ private static int encodedLengthGeneral(CharSequence sequence, int start)
   }
 
   /**
-   * Decodes the given UTF-8 encoded byte array slice into a {@link String}.
-   *
-   * @throws IllegalArgumentException if the input is not valid UTF-8.
+   * Decodes the given ASCII encoded byte array slice into a {@link String}.
    */
-  public static String decode(byte[] bytes, int index, int size)
+  public static String decodeASCII(byte[] bytes, int index, int size, TextBuffer textBuffer)
   {
-    // Bitwise OR combines the sign bits so any negative value fails the check.
-    if ((index | size | bytes.length - index - size) < 0)
-    {
-      throw new ArrayIndexOutOfBoundsException(
-          String.format("buffer length=%d, index=%d, size=%d", bytes.length, index, size));
-    }
-
     int offset = index;
     final int limit = offset + size;
 
-    // The longest possible resulting String is the same as the number of input bytes, when it is
-    // all ASCII. For other cases, this over-allocates and we will truncate in the end.
-    char[] resultArr = new char[size];
-    int resultPos = 0;
+    // Reuse buffers to avoid thrashing due to transient allocs.
+    char[] resultArr = null;
+    try
+    {
+      resultArr = textBuffer.getBuf(size);
+      int resultPos = 0;
+      while (offset < limit) {
+        resultArr[resultPos++] = (char) bytes[offset++];
+      }
+      return new String(resultArr, 0, size);
+    }
+    finally
+    {
+      textBuffer.returnBuf(resultArr);
+    }
+  }
 
-    while (offset < limit)
+  /**
+   * Decodes a long ASCII encoded byte source that spans multiple byte array chunks into a {@link String}.
+   */
+  public static String decodeLongASCII(LongDecoderState state, int size, TextBuffer textBuffer) throws IOException
+  {
+    // Reuse buffers to avoid thrashing due to transient allocs.
+    char[] resultArr = null;
+    try
     {
-      byte byte1 = bytes[offset++];
-      int value = byte1 & 0xff;
-      switch (UTF8_LOOKUP_TABLE[value])
+      resultArr = textBuffer.getBuf(size);
+      int resultPos = 0;
+
+      byte[] buffer = state._buffer;
+      int position = state._position;
+      int limit = state._offset + state._bufferSize;
```

Comment:
Consider moving limit inside the state. _offset and _bufferSize are not need outside of calculating limit.

Reply:
I prefer keeping it this way. This way we will have more flexibility in the future when dealing with incremental chunked data sources (if we need to like chrysaor does today).